datafusion/core/src/sql/planner.rs

```diff
@@ -1913,6 +1911,16 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         })
     }
 
+    fn parse_scalar_subquery(
+        &self,
+        subquery: &Query,
+        input_schema: &DFSchema,
+    ) -> Result<Expr> {
+        Ok(Expr::ScalarSubquery(Subquery {
+            subquery: Arc::new(self.subquery_to_plan(subquery.clone(), input_schema)?),
```

Comment:
So the scalar subquery must produce a single row typically otherwise it is a runtime error

Do you have thoughts about in what stage a query like the following would throw an error?

```
"SELECT p.id, (SELECT id FROM person) FROM person p"
```

Reply:
I was thinking that an optimizer rule should perform this type of validation check and this would cover both the SQL and DataFrame API (and maybe substrait one day) use cases.

Reply:
I filed https://github.com/apache/arrow-datafusion/issues/2361 to track the validation work